### PR TITLE
Corp api updates

### PR DESCRIPTION
**Pull request by pigalot**
Includes most required functions to make a corp from scratch.

[Test script](https://github.com/pigalot/bitburner-corp-api-demo)

Includes fixes for:
- Lack of validation on some corp UI/API (mostly negative numbers)
- Text on create Corp screen being wrong on non 3 bitnodes
- A bug were supply leftovers didn't remove spaces from material names causing an error.
- fixed a few random doc issues like the research function claiming to hire ads

Balance Related stuff:
- As discussed [here](https://discord.com/channels/415207508303544321/459097632896188436/930253820456489031) both corp api upgrades given permanently to play on completion of 3.3.
- Added a function that mimic the auto assign of employees like on the UI currently it waits for 1 second same as the assign employee, wasn't sure if this was a balance thing or if its just a legacy thing so decided it was best to match it. [Link](https://github.com/pigalot/bitburner/blob/38bd1276193f4f899ad2479c647ce91e26713c27/src/NetscriptFunctions/Corporation.ts#L531)

I have rebuilt the md docs, i can remove the commits for that if you have some other process for that.

### CODE REVIEW

File: src/Corporation/Industry.ts [hunk 1]
```diff
@@ -435,8 +435,17 @@ export class Industry implements IIndustry {
       const popularityGain = corporation.getDreamSenseGain(),
         awarenessGain = popularityGain * 4;
       if (popularityGain > 0) {
-        this.popularity += popularityGain * marketCycles;
-        this.awareness += awarenessGain * marketCycles;
+        if (this.awareness === Number.MAX_VALUE || this.awareness + (awarenessGain * marketCycles) > Number.MAX_VALUE) {
+          this.awareness == Number.MAX_VALUE;
```
[the remark this answers is not in the document]
Comment (pigalot) [reply src/Corporation/Industry.ts]: Agreed min is more elegant changed.

**Comment by phyzical** [comment]
Thanks @pigalot !! perfect timing for me :D just starting building a script looking for a bunch of this stuff.

File: markdown/bitburner.officeapi.md [hunk 2]
```diff
@@ -23,10 +23,15 @@ Requires the Office API upgrade from your corporation.
...
+|  [hasResearched(divisionName, researchName)](./bitburner.officeapi.hasresearched.md) | Gets if you have unlocked a research |
 |  [hireAdVert(divisionName)](./bitburner.officeapi.hireadvert.md) | Hire AdVert. |
-|  [hireEmployee(divisionName, cityName)](./bitburner.officeapi.hireemployee.md) | Assign an employee to a job. |
-|  [research(divisionName, researchName)](./bitburner.officeapi.research.md) | Hire AdVert. |
-|  [throwParty(divisionName, cityName, costPerEmployee)](./bitburner.officeapi.throwparty.md) | Assign an employee to a job. |
+|  [hireEmployee(divisionName, cityName)](./bitburner.officeapi.hireemployee.md) | Hire an employee. |
+|  [research(divisionName, researchName)](./bitburner.officeapi.research.md) | Purchase a research |
+|  [setAutoJobAssignment(divisionName, cityName, job, amount)](./bitburner.officeapi.setautojobassignment.md) | Set the auto job assignment for a job |
+|  [throwParty(divisionName, cityName, costPerEmployee)](./bitburner.officeapi.throwparty.md) | Throw a party for your employees |
 |  [upgradeOfficeSize(divisionName, cityName, size)](./bitburner.officeapi.upgradeofficesize.md) | Upgrade office size. |
```
Comment (phyzical) [inline markdown/bitburner.officeapi.md:36]: @pigalot one thing i've just noticed is missing, is there currently a way to get the cost for upgrading an office?

File: markdown/bitburner.product.developmentprogress.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
+<!-- Do not edit this file. It is automatically generated by API Documenter. -->
+
+[Home](./index.md) &gt; [bitburner](./bitburner.md) &gt; [Product](./bitburner.product.md) &gt; [developmentProgress](./bitburner.product.developmentprogress.md)
+
+## Product.developmentProgress property
+
+Creation progress - A number betwee 0-100 representing percentage
```
Comment (phyzical) [inline markdown/bitburner.product.developmentprogress.md:7]:
```suggestion
Creation progress - A number between 0-100 representing percentage
```

File: markdown/bitburner.product.md [hunk 4]
```diff
@@ -16,7 +16,9 @@ interface Product
 
 |  Property | Type | Description |
 |  --- | --- | --- |
+|  [cityData](./bitburner.product.citydata.md) | {\[key: string\]:number\[\]} | Data refers to the production, sale, and quantity of the products These values are specific to a city For each city, the data is \[qty, prod, sell\] |
 |  [cmp](./bitburner.product.cmp.md) | number | Competition for the product |
+|  [developmentProgress](./bitburner.product.developmentprogress.md) | number | Creation progress - A number betwee 0-100 representing percentage |
```
Comment (phyzical) [inline markdown/bitburner.product.md:21]:
```suggestion
|  [developmentProgress](./bitburner.product.developmentprogress.md) | number | Creation progress - A number between 0-100 representing percentage |
```

File: src/ScriptEditor/NetscriptDefinitions.d.ts [hunk 5]
```diff
@@ -6247,6 +6368,12 @@ interface Product {
   pCost: number;
   /** Sell cost, can be "MP+5" */
   sCost: string | number;
+  /** Data refers to the production, sale, and quantity of the products 
+   * These values are specific to a city 
+   * For each city, the data is [qty, prod, sell] */
+   cityData: {[key: string]:number[]};
+  /** Creation progress - A number betwee 0-100 representing percentage */
```
Comment (phyzical) [inline src/ScriptEditor/NetscriptDefinitions.d.ts:6375]:
```suggestion
  /** Creation progress - A number between 0-100 representing percentage */
```

File: markdown/bitburner.officeapi.md (hunk 2 above)
Comment (pigalot) [reply markdown/bitburner.officeapi.md:37]: Good catch, added.

**Comment by phyzical** [comment]
@pigalot thanks again :)

**Comment by phyzical** [comment]
@pigalot I also just stumbled across an unrelated bug where you can call `makeProduct` for an industry that shouldnt be able to have products i.e 'Chemical' and the product is created

**Comment by phyzical** [comment]
few more suggestions
* i feel is missing is access to the material boost modifiers for a division
image
* would it be possible to have a way to bulkPurchase materials? maybe an additional boolean to buyMaterial ?

**Comment by pigalot** [comment]
@phyzical I don't want this PR to keep going and going so I am trying to set some limits, its already rather large.

> * Access to the material boost modifiers for a division

This is hidden info as far as I am concerned so it can stay as it is.

> * would it be possible to have a way to bulkPurchase materials? maybe an additional boolean to buyMaterial ?
> * Ability to buyback shares

I'm not planning to add any more functions as of know, feel free to add a pr for it.

**Comment by phyzical** [comment]
@pigalot no worries :) thats fair enough.

i will look at implementing the later 2 a future pr 👍

**Pull request merged**